frontend/amundsen_application/static/js/components/Table/index.tsx

```diff
@@ -425,19 +544,82 @@ const getTableRows = (tableRowDetails: TableRowDetails) => {
       expandedRows.includes(item.key) &&
       formatChildrenData
     ) {
+      let formattedChildren = item.typeMetadata
+        ? getFormattedChildrenData(item, formatChildrenData)
+        : [];
+      let colToDisplay = item.typeMetadata ? formattedChildren[0] : item;
+
+      let arrayCount = 0;
+      while (colToDisplay.kind === ARRAY_KIND) {
+        // Keep track of how many nested array levels there are and skip its type metadata
+        arrayCount++;
+        formattedChildren = getFormattedChildrenData(
+          colToDisplay,
+          formatChildrenData
+        );
+
+        if (formattedChildren[0].isExpandable) {
+          [colToDisplay] = formattedChildren;
+        } else {
+          if (formattedChildren[0].kind === ARRAY_KIND) {
+            arrayCount++;
+          }
+          break;
+        }
+      }
+
+      if (colToDisplay.kind === ARRAY_KIND) {
+        // The innermost nested kind is array, so don't display an extra row for its terminal state
+        formattedChildren = [];
+      } else {
+        formattedChildren = getFormattedChildrenData(
+          colToDisplay,
+          formatChildrenData
+        );
+
+        if (colToDisplay.kind === MAP_KIND) {
+          formattedChildren = updateMapDisplayNames(formattedChildren);
+        }
+      }
+
+      const hasSpecificTypeHandling =
+        arrayCount > 0 || colToDisplay.kind === MAP_KIND;
+      const additionalTableColCount = columns.length - 1;
+
+      const openerRow = hasSpecificTypeHandling
+        ? [
+            getSpecificTypeOpenerRow(
+              colToDisplay,
+              arrayCount,
+              nestedLevel,
+              additionalTableColCount
+            ),
+          ]
+        : [];
+      const closerRow = hasSpecificTypeHandling
+        ? [
+            getSpecificTypeCloserRow(
+              colToDisplay,
+              arrayCount,
+              nestedLevel,
+              additionalTableColCount
+            ),
+          ]
+        : [];
```

Comment:
This was a bit too dense for me to read, could you try extracting some more functions?